src/main/java/net/neoforged/neoforge/transfer/handlers/wrappers/fluids/BlockFluidHandler.java

```diff
@@ -0,0 +1,197 @@
+        return FluidType.BUCKET_VOLUME;
+    }
+
+    @Override
+    public boolean isValid(int index, FluidResource resource) {
+        Objects.checkIndex(index, size());
+        return true;
+    }
+
+    @Override
+    public boolean supportsInsertion() {
+        return true;
+    }
+
+    @Override
+    public boolean supportsExtraction() {
+        return true;
+    }
+
+    @Override
+    public int insert(FluidResource resource, int amount, TransactionContext transaction) {
+        if (ResourceHandlerUtil.isEmpty(resource, amount)) return 0;
+        if (amount < FluidType.BUCKET_VOLUME) return 0;
+        //We already are handling an insert this callback, we can't do more until closing.
+        // Something like vaporizing needs to be deferred until commit, but we don't have a
+        // system that allows us to queue callbacks as first-in-first-out.
+        if (!fluidStack.isEmpty()) return 0;
+
+        Fluid fluid = resource.getInstanceValue();
+        FluidStack fluidstack = resource.toStack(amount);
+        if (!fluid.getFluidType().canBePlacedInLevel(level, pos, fluidstack)) return 0;
+
+        BlockState state = level.getBlockState(pos);
+
+        boolean isFluidContainer = state.getBlock() instanceof LiquidBlockContainer container && container.canPlaceLiquid(null, level, pos, state, resource.getInstanceValue());
+        boolean isReplaceable = state.canBeReplaced(resource.getInstanceValue());
+
+        if (!isFluidContainer && !isReplaceable) return 0;
+        updateSnapshots(transaction);
+        fluidStack = resource.toStack(amount);
+
+        if (resource.getFluidType().isVaporizedOnPlacement(level, pos, fluidstack)) {
+            //Handled in commit instead, though it should be noted: this likely only handles the LAST fluid inserted. Something to explore a solution to
+            //resource.getFluidType().onVaporize(player, level, pos, fluidstack);
+        } else if (isFluidContainer) {
+            //not a lot of choice during transaction. We must defer to the implementer's choice of setting block
+            //TODO how should we handle this. If someone has an implementation of this class they'd like to propose, we can try it.
+            ((LiquidBlockContainer) state.getBlock()).placeLiquid(level, pos, state, resource.getInstanceValue().defaultFluidState());
+        } else {
+            //defer breaking until commit
+//            FluidUtil.destroyBlockOnFluidPlacement(level, pos);
+            level.setBlock(pos, resource.getInstanceValue().defaultFluidState().createLegacyBlock(), 0);
+        }
+
+        return FluidType.BUCKET_VOLUME;
+    }
+
+    protected record Snapshot(BlockState state, FluidStack fluid) {}
+
+    @Override
+    public int extract(FluidResource resource, int amount, TransactionContext transaction) {
+        if (ResourceHandlerUtil.isEmpty(resource, amount)) return 0;
+        //We already are handling an insert this callback, we can't do more until closing.
```

Comment:
Is this meant to say insert? Given this is in the extract method? My guess is yes, but I just wanted to make sure